File: Bridge/System/Array.cs
using Bridge;
using System.Collections;
using System.Collections.Generic;

namespace System
{
    [External]
    [Namespace("Bridge")]
    public sealed class Array : IEnumerable, ICloneable
    {
        [Template("#{this}")]
        public readonly int Length = 0;

        private Array()
        {
        }

        public extern IEnumerator GetEnumerator();

        public extern object GetValue(params int[] indices);

        public extern void SetValue(object value, params int[] indices);

        public extern int GetLength(int dimension);

        public int Rank
        {
            get
            {
                return 0;
            }
        }

        public object Clone() {
            return null;
        }

        #region array

        public static T[] Empty<T>() {
            return null;
        }

        public static bool Exists<T>(T[] array, Predicate<T> match) {
            return false;
        }

        public static T Find<T>(T[] array, Predicate<T> match) {
            return default(T);
        }

        public static T[] FindAll<T>(T[] array, Predicate<T> match) {
            return null;
        }

        public static int FindIndex<T>(T[] array, Predicate<T> match) {
            return 0;
        }

        public static int FindIndex<T>(T[] array, int startIndex, Predicate<T> match) {
            return 0;
        }

        public static int FindIndex<T>(T[] array, int startIndex, int count, Predicate<T> match) {
            return 0;
        }

        public static T FindLast<T>(T[] array, Predicate<T> match) {
            return default(T);
        }

        public static int FindLastIndex<T>(T[] array, Predicate<T> match) {
            return 0;
        }

        public static int FindLastIndex<T>(T[] array, int startIndex, Predicate<T> match) {
            return 0;
        }

        public static int FindLastIndex<T>(T[] array, int startIndex, int count, Predicate<T> match) {
            return 0;
        }

        public static void ForEach<T>(T[] array, Action<T> action) {
        }

        public static int IndexOf<T>(T[] array, T value) {
            return 0;
        }

        public static int IndexOf<T>(T[] array, T value, int startIndex) {
            return 0;
        }

        public static int IndexOf<T>(T[] array, T value, int startIndex, int count) {
            return 0;
        }

        public static int LastIndexOf<T>(T[] array, T value) {
            return 0;
        }

        public static int LastIndexOf<T>(T[] array, T value, int startIndex) {
            return 0;
        }

        public static int LastIndexOf<T>(T[] array, T value, int startIndex, int count) {
            return 0;
        }

        public static void Reverse<T>(T[] array) {
            Array.Reverse(array);
        }

        public static void Reverse<T>(T[] array, int index, int length) {
        }

        public static void Sort<T>(T[] array) { }

        public static void Sort<T>(T[] array, IComparer<T> comparer) { }

        public static void Sort<T>(T[] array, Comparison<T> comparison) {
        }

        public static void Sort<T>(T[] array, int index, int length, IComparer<T> comparer) { }

        public static void Sort<T>(T[] array, int index, int length) { }

        public static bool TrueForAll<T>(T[] array, Predicate<T> match) {
            return false;
        }


        #endregion
    }
}

File: Compiler/Lua/CoreSystem.Lua/CoreSystem/Array.lua
local System = System
local Collection = System.Collection
local buildArray = Collection.buildArray
local checkIndex = Collection.checkIndex 
local arrayEnumerator = Collection.arrayEnumerator

local Array = {}
local emptys = {}

Array.__ctor__ = buildArray
Array.set = Collection.setArray
Array.get = Collection.getArray
Array.getEnumerator = arrayEnumerator

function Array.getLength(this)
    return #this
end

function Array.getRank(this)
   return 1
end

function Array.Empty(T)
    local t = emptys[T]
    if t == nil then
        t = Array(T)(0)
        emptys[T] = t
    end
    return t
end

Array.exists = Collection.existsOfArray
Array.find = Collection.findOfArray

local findAll = Collection.findAllOfArray
function Array.findAll(t, match)
    return findAll(t, match):toArray()
end

Array.findIndex = Collection.findIndexOfArray
Array.findLast = Collection.findLastOfArray
Array.findLastIndex = Collection.findLastIndexOfArray
Array.indexOf = Collection.indexOfArray
Array.lastIndexOf = Collection.lastIndexOfArray
Array.reverse = Collection.reverseArray
Array.sort = Collection.sortArray
Array.trueForAll = Collection.trueForAllOfArray
Array.copy = Collection.copyArray

System.define("System.Array", function(T) 
    local cls = { 
    __inherits__ = { System.ICollection_1(T), System.ICollection, System.IList_1(T) }, 
    __genericT__ = T
    }
    return cls
end, Array)

function System.arrayFromTable(t, T)
    return setmetatable(t, System.Array(T))
end

local MultiArray = {}

function MultiArray.__ctor__(this, rank, ...)
    this.__rank__ = rank
    local length = 1
    for _, i in ipairs(rank) do
        length = length * i
    end
    buildArray(this, length, ...)
end

local function getIndex(this, ...)
    local rank = this.__rank__
    local id = 0
    local len = #rank
    for i = 1, len do
        id = id * rank[i] + select(i, ...)
    end
    return id, len
end

function MultiArray.set(this, ...)
    local index, len = getIndex(this, ...)
    setArray(this, index, select(len + 1, ...))
end

function MultiArray.get(this, ...)
    local index = getIndex(this, ...)
    return getArray(this, index)
end

function MultiArray.getLength(this, dimension)
    if dimension == nil then
        return #this
    end
    local rank = thiss.__rank__
    checkIndex(rank, dimension)
    return this.__rank__[dimension + 1]
end

function MultiArray.getRank(this)
   return #this.__rank__
end

MultiArray.getEnumerator = arrayEnumerator

System.define("System.MultiArray", function(T) 
    local cls = { 
    __inherits__ = { System.ICollection_1(T), System.ICollection, System.IList_1(T) }, 
    __genericT__ = T
    }
    return cls
end, MultiArray)